woocommerce/woocommerce
# Basic range download handling for Force download.

### All Submissions:






### Changes proposed in this Pull Request:



This change adds support for HTTP_RANGE requests (used by iOS devices) to `Force download` method.
One thing to note is that range request downloads are not logged (so that they are not counted as multiple downloads). The most reasonable option in my view would be to add extra column(s) to table `wp_wc_download_log`, e.g. columns `start` and `length` and then exclude range request log entries when counting number of downloads. 
Not sure how feasible db changes are mid-release-cycle, so I went for easier option for now--ignoring ranged downloads from the log. 
As the initial request from an iOS device is a normal (non-range) request, the download should be logged once anyway.
 
Closes #19640 .

### How to test the changes in this Pull Request:

1. Create a downloadable virtual product which contains larger audio or video file, e.g. an mp3 file.
2. Buy the product, go to My account > Downloads.
3. Try to download the mp3 file on iPhone/iPad.
4. Audio only plays first ~2 minutes, then starts from the beginning again, it does not indicate the length and does not allow seeking.
5. Apply this branch.
6. See the audio now shows correct length and allows seeking.

### Other information:

* [ ] Have you written new tests for your changes, as applicable? No, it's a bit tricky as headers interfere with phpunit's output and e2e tests don't run on iOS device.



### Changelog entry

> Added support for HTTP_RANGE requests (used by iOS devices) to `Force download` method.
